### PR TITLE
wal: atomically initialize wal directory

### LINKED ISSUE #5270: etcdserver: panics on RestartNode if WAL partially initialized
Got this from burning etcd with local-tester using a -race build.

Looks like this happens if etcd is killed between creating its first WAL file and syncing it down since the wal file is 0 bytes and etcd3 never fully joined the cluster.

```
11:03:02      etcd3 | 2016-05-04 11:03:02.936918 I | etcdmain: etcd Version: 2.3.0+git
11:03:02      etcd3 | 2016-05-04 11:03:02.937076 I | etcdmain: Git SHA: 7ea955a
11:03:02      etcd3 | 2016-05-04 11:03:02.937107 I | etcdmain: Go Version: go1.6
11:03:02      etcd3 | 2016-05-04 11:03:02.937139 I | etcdmain: Go OS/Arch: linux/amd64
11:03:02      etcd3 | 2016-05-04 11:03:02.937176 I | etcdmain: setting maximum number of CPUs to 8, total number of available CPUs is 8
11:03:02      etcd3 | 2016-05-04 11:03:02.937218 W | etcdmain: no data-dir provided, using default data-dir ./infra3.etcd
11:03:02      etcd3 | 2016-05-04 11:03:02.937277 N | etcdmain: the server is already initialized as member before, starting as etcd member...
11:03:02      etcd3 | 2016-05-04 11:03:02.937383 I | etcdmain: listening for peers on http://127.0.0.1:32380
11:03:02      etcd3 | 2016-05-04 11:03:02.937447 I | etcdmain: listening for client requests on 127.0.0.1:33339
11:03:02      etcd3 | 2016-05-04 11:03:02.937939 I | etcdserver: name = infra3
11:03:02      etcd3 | 2016-05-04 11:03:02.937978 I | etcdserver: data dir = infra3.etcd
11:03:02      etcd3 | 2016-05-04 11:03:02.938014 I | etcdserver: member dir = infra3.etcd/member
11:03:02      etcd3 | 2016-05-04 11:03:02.938049 I | etcdserver: heartbeat = 100ms
11:03:02      etcd3 | 2016-05-04 11:03:02.938081 I | etcdserver: election = 1000ms
11:03:02      etcd3 | 2016-05-04 11:03:02.938112 I | etcdserver: snapshot count = 1000
11:03:02      etcd3 | 2016-05-04 11:03:02.938152 I | etcdserver: advertise client URLs = http://127.0.0.1:32379
11:03:02      etcd3 | 2016-05-04 11:03:02.952918 I | etcdserver: restarting member 0 in cluster 0 at commit index 0
11:03:02      etcd3 | panic:
11:03:02      etcd3 | cannot use none as id
11:03:02      etcd3 | goroutine 1 [running]:
11:03:02      etcd3 | panic(0xf54d40, 0xc8202431e0)
11:03:02      etcd3 |   /usr/lib/go/src/runtime/panic.go:464 +0x3ff
11:03:02      etcd3 | github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.newRaft(0xc82017a018, 0x43d1be)
11:03:02      etcd3 |   /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/raft.go:191 +0x1d7
11:03:02      etcd3 | github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.RestartNode(0xc82017a018, 0x0, 0x0)
11:03:02      etcd3 |   /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/node.go:212 +0x56
11:03:02      etcd3 | github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver.restartNode(0xc82000a780, 0x0, 0x7efc4cbb1000, 0xc820015100, 0x0, 0x0, 0x0, 0x0)
11:03:02      etcd3 |   /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/raft.go:349 +0xabc
11:03:02      etcd3 | github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver.NewServer(0xc82000a780, 0x0, 0x7efc4cbb1000, 0xc820015100)
11:03:02      etcd3 |   /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/server.go:348 +0x4d36
11:03:02      etcd3 | github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.startEtcd(
11:03:02      etcd3 | 0xc82001a780, 0x0, 0x0, 0x0)
11:03:02      etcd3 |   /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/etcd.go:369 +0x33ca
11:03:02      etcd3 | github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.Main()
11:03:02      etcd3 |   /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/etcd.go:118 +0x290b
11:03:02      etcd3 | main.main()
11:03:02      etcd3 |   /home/anthony/go/src/github.com/coreos/etcd/cmd/main.go:28 +0x26
11:03:02      etcd3 | Terminating etcd3
```

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #5270

### CODE REVIEW

**Comment by xiang90** [comment]
LGTM

File: wal/wal_test.go [hunk 1]
```diff
@@ -600,3 +600,39 @@ func TestTailWriteNoSlackSpace(t *testing.T) {
...
+		t.Fatalf("got %q exists, expected it to not exist", tmpdir)
+	}
+
+	if w, err = OpenForRead(p, walpb.Snapshot{}); err != nil {
+		t.Fatal(err)
+	}
+	defer w.Close()
+
+	if meta, _, _, rerr := w.ReadAll(); rerr != nil || string(meta) != "abc" {
+		t.Fatalf("got error %v and meta %q, expected nil and %q", rerr, string(meta), "abc")
```
Comment (gyuho) [inline wal/wal_test.go]: we don't need `string(` because format string is already `%q`?

**Comment by gyuho** [comment]
LGTM. Thanks.

File: wal/wal_test.go (hunk 1 above)
Comment (heyitsanthony) [inline wal/wal_test.go]: you're right-- will fix

**Pull request merged**